File: backend/services/motor_paridad.py
import math
from collections import defaultdict

from backend.constants.legislacion import (
    ANIO_JUVENTUD_MAX,
    ANIO_JUVENTUD_MIN,
    ART_ACCION_DISCAPACIDAD,
    ART_ACCION_INDIGENA,
    ART_ACCION_JUVENTUD,
    ART_ACCION_LGBTQ,
    ART_PARIDAD_HORIZONTAL,
    ART_PARIDAD_TRANSVERSAL,
    ART_PARIDAD_VERTICAL,
    UMBRAL_DISCAPACIDAD,
    UMBRAL_INDIGENA,
    UMBRAL_JUVENTUD,
    UMBRAL_LGBTQ,
    UMBRAL_TRANSVERSAL_TOLERANCIA,
)
from backend.models.candidatura import Candidatura, Incumplimiento, ResultadoValidacion


# ---------------------------------------------------------------------------
# Regla 1 — Paridad Horizontal (Art. 234 LGIPE)
# ---------------------------------------------------------------------------

def _paridad_horizontal(candidaturas: list[Candidatura]) -> tuple[dict, list[Incumplimiento]]:
    """
    Verifica que en cada fórmula propietario-suplente los géneros sean distintos.
    Solo aplica a pares M/F; si alguno es NB la fórmula queda exenta del check binario.
    Agrupa por (posicion, distrito, cargo).
    """
    formulas: dict[tuple, dict[str, Candidatura]] = defaultdict(dict)
    for c in candidaturas:
        llave = (c.posicion, c.distrito, c.cargo)
        formulas[llave][c.tipo] = c

    total_formulas = 0
    violaciones: list[Incumplimiento] = []

    for llave, miembros in formulas.items():
        propietario = miembros.get("PROPIETARIO")
        suplente    = miembros.get("SUPLENTE")

        if not propietario or not suplente:
            # Fórmula incompleta — no se valida paridad horizontal aquí
            continue

        total_formulas += 1
        pos, distrito, cargo = llave

        # Exentar fórmulas donde algún miembro es NB
        if propietario.genero == "NB" or suplente.genero == "NB":
            continue

        if propietario.genero == suplente.genero:
            violaciones.append(
                Incumplimiento(
                    tipo="PARIDAD_HORIZONTAL",
                    descripcion=(
                        f"La fórmula en posición {pos}, distrito '{distrito}', cargo '{cargo}' "
                        f"tiene propietario/a y suplente del mismo género ({propietario.genero})."
                    ),
                    candidatos_afectados=[propietario.nombre, suplente.nombre],
                    articulo=ART_PARIDAD_HORIZONTAL,
                    sugerencia=(
                        f"Cambiar el género del/la suplente '{suplente.nombre}' en posición {pos}, "
                        f"distrito '{distrito}' para que sea distinto al del/la propietario/a."
                    ),
                )
            )

    formulas_validas = total_formulas - len(violaciones)
    pct = (formulas_validas / total_formulas * 100) if total_formulas > 0 else 100.0

    criterio = {
        "cumple": len(violaciones) == 0,
        "porcentaje_cumplimiento": round(pct, 2),
        "articulo": ART_PARIDAD_HORIZONTAL,
        "descripcion": (
            "Cada fórmula propietario-suplente debe tener géneros distintos (M/F). "
            "Candidaturas con género NB están exentas del check binario."
        ),
        "formulas_revisadas": total_formulas,
        "formulas_con_violacion": len(violaciones),
    }
    return criterio, violaciones


# ---------------------------------------------------------------------------
# Regla 2 — Paridad Vertical (Art. 233 LGIPE)
# ---------------------------------------------------------------------------

def _paridad_vertical(candidaturas: list[Candidatura]) -> tuple[dict, list[Incumplimiento]]:
    """
    Verifica alternancia de géneros en listas ordenadas.
    Agrupa por (partido, cargo, tipo), ordena por posicion, aplica ventana deslizante de 2.
    IMPORTANTE: propietarios y suplentes se evalúan por separado — la alternancia
    de Art. 233 aplica a la lista ordenada de propietarios (y separadamente a la
    de suplentes), no a la mezcla de ambos.
    Candidaturas NB se excluyen del análisis de alternancia binaria.
    """
    listas: dict[tuple, list[Candidatura]] = defaultdict(list)
    for c in candidaturas:
        listas[(c.partido, c.cargo, c.tipo)].append(c)

    total_pares = 0
    violaciones: list[Incumplimiento] = []

    for (partido, cargo, tipo), lista in listas.items():
        # Solo M y F participan en la alternancia
        binarios = sorted(
            [c for c in lista if c.genero in ("M", "F")],
            key=lambda x: x.posicion,
        )

        if len(binarios) < 2:
            continue

        for i in range(len(binarios) - 1):
            total_pares += 1
            a = binarios[i]
            b = binarios[i + 1]
            if a.genero == b.genero:
                violaciones.append(
                    Incumplimiento(
                        tipo="PARIDAD_VERTICAL",
                        descripcion=(
                            f"Posiciones consecutivas {a.posicion} y {b.posicion} del cargo "
                            f"'{cargo}' (partido '{partido}', tipo '{tipo}') tienen el "
                            f"mismo género ({a.genero})."
                        ),
                        candidatos_afectados=[a.nombre, b.nombre],
                        articulo=ART_PARIDAD_VERTICAL,
                        sugerencia=(
                            f"Intercambiar la posición {b.posicion} ('{b.nombre}', {b.genero}) "
                            f"con alguna candidatura {tipo} de género distinto para lograr alternancia."
                        ),
                    )
                )

    pares_validos = total_pares - len(violaciones)
    pct = (pares_validos / total_pares * 100) if total_pares > 0 else 100.0

    criterio = {
        "cumple": len(violaciones) == 0,
        "porcentaje_cumplimiento": round(pct, 2),
        "articulo": ART_PARIDAD_VERTICAL,
        "descripcion": (
            "Las listas de candidaturas ordenadas deben alternar géneros M/F. "
            "Se evalúa con ventana deslizante de 2 posiciones consecutivas."
        ),
        "pares_revisados": total_pares,
        "pares_con_violacion": len(violaciones),
    }
    return criterio, violaciones


# ---------------------------------------------------------------------------
# Regla 3 — Paridad Transversal (Art. 232 LGIPE)
# ---------------------------------------------------------------------------

def _paridad_transversal(candidaturas: list[Candidatura]) -> tuple[dict, list[Incumplimiento]]:
    """
    Verifica que el total de candidaturas sea 50% M y 50% F (±5%).
    Candidaturas NB se excluyen del cálculo binario.
    """
    hombres = sum(1 for c in candidaturas if c.genero == "M")
    mujeres  = sum(1 for c in candidaturas if c.genero == "F")
    total_binarios = hombres + mujeres

    if total_binarios == 0:
        criterio = {
            "cumple": False,
            "porcentaje_cumplimiento": 0.0,
            "articulo": ART_PARIDAD_TRANSVERSAL,
            "descripcion": "No hay candidaturas con género M o F para evaluar paridad transversal.",
            "mujeres": 0,
            "hombres": 0,
        }
        return criterio, []

    pct_f = mujeres / total_binarios
    cumple = abs(pct_f - 0.5) <= UMBRAL_TRANSVERSAL_TOLERANCIA

    # Porcentaje de cumplimiento: 100% = exactamente 50/50; decrece si se aleja
    pct_cumplimiento = max(0.0, 100.0 - abs(pct_f - 0.5) * 200)

    incumplimientos: list[Incumplimiento] = []
    if not cumple:
        genero_minoritario = "F" if mujeres < hombres else "M"
        delta = math.ceil(abs(hombres - mujeres) / 2)
        incumplimientos.append(
            Incumplimiento(
                tipo="PARIDAD_TRANSVERSAL",
                descripcion=(
                    f"Distribución actual: {mujeres} mujeres ({pct_f*100:.1f}%) y "
                    f"{hombres} hombres ({(1-pct_f)*100:.1f}%). "
                    f"Se requiere 50% ±{UMBRAL_TRANSVERSAL_TOLERANCIA*100:.0f}%."
                ),
                candidatos_afectados=[],
                articulo=ART_PARIDAD_TRANSVERSAL,
                sugerencia=(
                    f"Agregar o sustituir al menos {delta} candidatura(s) de género "
                    f"'{genero_minoritario}' para alcanzar la paridad del 50% (tolerancia ±5%)."
                ),
            )
        )

    criterio = {
        "cumple": cumple,
        "porcentaje_cumplimiento": round(pct_cumplimiento, 2),
        "articulo": ART_PARIDAD_TRANSVERSAL,
        "descripcion": (
            "El total de candidaturas debe ser 50% mujeres y 50% hombres, con tolerancia del ±5%."
        ),
        "mujeres": mujeres,
        "hombres": hombres,
        "porcentaje_mujeres": round(pct_f * 100, 2),
    }
    return criterio, incumplimientos


# ---------------------------------------------------------------------------
# Reglas 4–7 — Acciones Afirmativas (Art. 14 LGIPE)
# ---------------------------------------------------------------------------

def _accion_afirmativa(
    candidaturas: list[Candidatura],
    campo: str,
    umbral: float,
    articulo: str,
    nombre_accion: str,
    tipo_incumplimiento: str,
    descripcion_criterio: str,
) -> tuple[dict, list[Incumplimiento]]:
    """Plantilla genérica para las 4 acciones afirmativas."""
    total = len(candidaturas)
    registrado = sum(1 for c in candidaturas if getattr(c, campo))
    requerido = math.ceil(total * umbral)

    cumple = registrado >= requerido
    pct = min(100.0, (registrado / requerido * 100) if requerido > 0 else 100.0)

    incumplimientos: list[Incumplimiento] = []
    if not cumple:
        faltantes = requerido - registrado
        incumplimientos.append(
            Incumplimiento(
                tipo=tipo_incumplimiento,
                descripcion=(
                    f"Se registraron {registrado} candidatura(s) de {nombre_accion}, "
                    f"pero se requieren al menos {requerido} "
                    f"({umbral*100:.0f}% de {total} candidaturas)."
                ),
                candidatos_afectados=[],
                articulo=articulo,
                sugerencia=(
                    f"Registrar al menos {faltantes} candidatura(s) adicional(es) "
                    f"de {nombre_accion}."
                ),
            )
        )

    criterio = {
        "cumple": cumple,
        "porcentaje_cumplimiento": round(pct, 2),
        "articulo": articulo,
        "descripcion": descripcion_criterio,
        "registrado": registrado,
        "requerido": requerido,
    }
    return criterio, incumplimientos


def _accion_juventud(candidaturas: list[Candidatura]) -> tuple[dict, list[Incumplimiento]]:
    """Acción afirmativa de juventud: nacidos entre ANIO_JUVENTUD_MIN y ANIO_JUVENTUD_MAX."""
    total = len(candidaturas)
    registrado = sum(
        1 for c in candidaturas
        if ANIO_JUVENTUD_MIN <= c.fecha_nacimiento.year <= ANIO_JUVENTUD_MAX
    )
    requerido = math.ceil(total * UMBRAL_JUVENTUD)
    cumple = registrado >= requerido
    pct = min(100.0, (registrado / requerido * 100) if requerido > 0 else 100.0)

    incumplimientos: list[Incumplimiento] = []
    if not cumple:
        faltantes = requerido - registrado
        incumplimientos.append(
            Incumplimiento(
                tipo="ACCION_JUVENTUD",
                descripcion=(
                    f"Se registraron {registrado} candidatura(s) de personas jóvenes "
                    f"(nacidas {ANIO_JUVENTUD_MIN}-{ANIO_JUVENTUD_MAX}), "
                    f"pero se requieren al menos {requerido} "
                    f"({UMBRAL_JUVENTUD*100:.0f}% de {total} candidaturas)."
                ),
                candidatos_afectados=[],
                articulo=ART_ACCION_JUVENTUD,
                sugerencia=(
                    f"Registrar al menos {faltantes} candidatura(s) adicional(es) "
                    f"de personas nacidas entre {ANIO_JUVENTUD_MIN} y {ANIO_JUVENTUD_MAX}."
                ),
            )
        )

    criterio = {
        "cumple": cumple,
        "porcentaje_cumplimiento": round(pct, 2),
        "articulo": ART_ACCION_JUVENTUD,
        "descripcion": (
            f"Al menos {UMBRAL_JUVENTUD*100:.0f}% de las candidaturas deben ser de personas "
            f"jóvenes (nacidas entre {ANIO_JUVENTUD_MIN} y {ANIO_JUVENTUD_MAX})."
        ),
        "registrado": registrado,
        "requerido": requerido,
    }
    return criterio, incumplimientos


# ---------------------------------------------------------------------------
# Función principal
# ---------------------------------------------------------------------------

def validar(candidaturas: list[Candidatura]) -> ResultadoValidacion:
    """
    Aplica las 7 reglas de paridad y acciones afirmativas.
    Devuelve ResultadoValidacion con criterios, incumplimientos y semáforo global.
    """
    partido = candidaturas[0].partido if candidaturas else "DESCONOCIDO"
    total   = len(candidaturas)

    # Aplicar las 7 reglas
    c_horizontal,    v_horizontal    = _paridad_horizontal(candidaturas)
    c_vertical,      v_vertical      = _paridad_vertical(candidaturas)
    c_transversal,   v_transversal   = _paridad_transversal(candidaturas)

    c_indigena, v_indigena = _accion_afirmativa(
        candidaturas, "indigena", UMBRAL_INDIGENA, ART_ACCION_INDIGENA,
        "personas indígenas", "ACCION_INDIGENA",
        f"Al menos {UMBRAL_INDIGENA*100:.0f}% de las candidaturas deben ser de personas indígenas.",
    )
    c_discapacidad, v_discapacidad = _accion_afirmativa(
        candidaturas, "discapacidad", UMBRAL_DISCAPACIDAD, ART_ACCION_DISCAPACIDAD,
        "personas con discapacidad", "ACCION_DISCAPACIDAD",
        f"Al menos {UMBRAL_DISCAPACIDAD*100:.0f}% de las candidaturas deben ser de personas con discapacidad.",
    )
    c_juventud, v_juventud = _accion_juventud(candidaturas)
    c_lgbtq, v_lgbtq = _accion_afirmativa(
        candidaturas, "lgbtq", UMBRAL_LGBTQ, ART_ACCION_LGBTQ,
        "personas LGBTQ+", "ACCION_LGBTQ",
        f"Al menos {UMBRAL_LGBTQ*100:.0f}% de las candidaturas deben ser de personas LGBTQ+.",
    )

    # Semáforo global
    falla_paridad = (
        not c_horizontal["cumple"]
        or not c_vertical["cumple"]
        or not c_transversal["cumple"]
    )
    falla_afirmativa = (
        not c_indigena["cumple"]
        or not c_discapacidad["cumple"]
        or not c_juventud["cumple"]
        or not c_lgbtq["cumple"]
    )

    if falla_paridad:
        resultado_global = "DANGER"
    elif falla_afirmativa:
        resultado_global = "WARNING"
    else:
        resultado_global = "SUCCESS"

    todos_incumplimientos = (
        v_horizontal + v_vertical + v_transversal
        + v_indigena + v_discapacidad + v_juventud + v_lgbtq
    )

    return ResultadoValidacion(
        partido=partido,
        total_candidaturas=total,
        resultado_global=resultado_global,
        criterios={
            "paridad_horizontal":  c_horizontal,
            "paridad_vertical":    c_vertical,
            "paridad_transversal": c_transversal,
            "accion_indigena":     c_indigena,
            "accion_discapacidad": c_discapacidad,
            "accion_juventud":     c_juventud,
            "accion_lgbtq":        c_lgbtq,
        },
        incumplimientos=todos_incumplimientos,
    )

File: frontend/src/demo/motorParidad.js
// Port a JavaScript del motor de reglas de `backend/services/motor_paridad.py`.
// Se usa SOLO en modo demo (GitHub Pages), cuando no hay backend FastAPI.
// La lógica y los artículos legales replican 1:1 la implementación de Python;
// si cambias una regla en el backend, refléjala aquí.

// --- Artículos legales (backend/constants/legislacion.py) ---
const ART_PARIDAD_HORIZONTAL = 'Art. 234 LGIPE'
const ART_PARIDAD_VERTICAL = 'Art. 233 LGIPE'
const ART_PARIDAD_TRANSVERSAL = 'Art. 232 LGIPE'
const ART_ACCION = 'Art. 14 LGIPE'

// --- Umbrales numéricos ---
const UMBRAL_TRANSVERSAL_TOLERANCIA = 0.05
const UMBRAL_INDIGENA = 0.03
const UMBRAL_DISCAPACIDAD = 0.01
const UMBRAL_JUVENTUD = 0.01
const UMBRAL_LGBTQ = 0.01
const ANIO_JUVENTUD_MIN = 1997
const ANIO_JUVENTUD_MAX = 2008

const round2 = (x) => Math.round((x + Number.EPSILON) * 100) / 100
const anioNacimiento = (c) => parseInt(String(c.fecha_nacimiento).slice(0, 4), 10)

// ---------------------------------------------------------------------------
// Regla 1 — Paridad Horizontal (Art. 234 LGIPE)
// ---------------------------------------------------------------------------
function paridadHorizontal(candidaturas) {
  const formulas = new Map() // (posicion|distrito|cargo) -> { PROPIETARIO, SUPLENTE }
  for (const c of candidaturas) {
    const llave = `${c.posicion}|${c.distrito}|${c.cargo}`
    if (!formulas.has(llave)) formulas.set(llave, {})
    formulas.get(llave)[c.tipo] = c
  }

  let totalFormulas = 0
  const violaciones = []

  for (const [llave, miembros] of formulas) {
    const propietario = miembros.PROPIETARIO
    const suplente = miembros.SUPLENTE
    if (!propietario || !suplente) continue // fórmula incompleta

    totalFormulas += 1
    const [pos, distrito, cargo] = llave.split('|')

    if (propietario.genero === 'NB' || suplente.genero === 'NB') continue

    if (propietario.genero === suplente.genero) {
      violaciones.push({
        tipo: 'PARIDAD_HORIZONTAL',
        descripcion:
          `La fórmula en posición ${pos}, distrito '${distrito}', cargo '${cargo}' ` +
          `tiene propietario/a y suplente del mismo género (${propietario.genero}).`,
        candidatos_afectados: [propietario.nombre, suplente.nombre],
        articulo: ART_PARIDAD_HORIZONTAL,
        sugerencia:
          `Cambiar el género del/la suplente '${suplente.nombre}' en posición ${pos}, ` +
          `distrito '${distrito}' para que sea distinto al del/la propietario/a.`,
      })
    }
  }

  const formulasValidas = totalFormulas - violaciones.length
  const pct = totalFormulas > 0 ? (formulasValidas / totalFormulas) * 100 : 100.0

  const criterio = {
    cumple: violaciones.length === 0,
    porcentaje_cumplimiento: round2(pct),
    articulo: ART_PARIDAD_HORIZONTAL,
    descripcion:
      'Cada fórmula propietario-suplente debe tener géneros distintos (M/F). ' +
      'Candidaturas con género NB están exentas del check binario.',
    formulas_revisadas: totalFormulas,
    formulas_con_violacion: violaciones.length,
  }
  return [criterio, violaciones]
}

// ---------------------------------------------------------------------------
// Regla 2 — Paridad Vertical (Art. 233 LGIPE)
// ---------------------------------------------------------------------------
function paridadVertical(candidaturas) {
  const listas = new Map() // (partido|cargo|tipo) -> Candidatura[]
  for (const c of candidaturas) {
    const llave = `${c.partido}|${c.cargo}|${c.tipo}`
    if (!listas.has(llave)) listas.set(llave, [])
    listas.get(llave).push(c)
  }

  let totalPares = 0
  const violaciones = []

  for (const [llave, lista] of listas) {
    const binarios = lista
      .filter((c) => c.genero === 'M' || c.genero === 'F')
      .sort((a, b) => a.posicion - b.posicion)

    if (binarios.length < 2) continue

    const [partido, cargo, tipo] = llave.split('|')
    for (let i = 0; i < binarios.length - 1; i += 1) {
      totalPares += 1
      const a = binarios[i]
      const b = binarios[i + 1]
      if (a.genero === b.genero) {
        violaciones.push({
          tipo: 'PARIDAD_VERTICAL',
          descripcion:
            `Posiciones consecutivas ${a.posicion} y ${b.posicion} del cargo ` +
            `'${cargo}' (partido '${partido}', tipo '${tipo}') tienen el ` +
            `mismo género (${a.genero}).`,
          candidatos_afectados: [a.nombre, b.nombre],
          articulo: ART_PARIDAD_VERTICAL,
          sugerencia:
            `Intercambiar la posición ${b.posicion} ('${b.nombre}', ${b.genero}) ` +
            `con alguna candidatura ${tipo} de género distinto para lograr alternancia.`,
        })
      }
    }
  }

  const paresValidos = totalPares - violaciones.length
  const pct = totalPares > 0 ? (paresValidos / totalPares) * 100 : 100.0

  const criterio = {
    cumple: violaciones.length === 0,
    porcentaje_cumplimiento: round2(pct),
    articulo: ART_PARIDAD_VERTICAL,
    descripcion:
      'Las listas de candidaturas ordenadas deben alternar géneros M/F. ' +
      'Se evalúa con ventana deslizante de 2 posiciones consecutivas.',
    pares_revisados: totalPares,
    pares_con_violacion: violaciones.length,
  }
  return [criterio, violaciones]
}

// ---------------------------------------------------------------------------
// Regla 3 — Paridad Transversal (Art. 232 LGIPE)
// ---------------------------------------------------------------------------
function paridadTransversal(candidaturas) {
  const hombres = candidaturas.filter((c) => c.genero === 'M').length
  const mujeres = candidaturas.filter((c) => c.genero === 'F').length
  const totalBinarios = hombres + mujeres

  if (totalBinarios === 0) {
    return [
      {
        cumple: false,
        porcentaje_cumplimiento: 0.0,
        articulo: ART_PARIDAD_TRANSVERSAL,
        descripcion: 'No hay candidaturas con género M o F para evaluar paridad transversal.',
        mujeres: 0,
        hombres: 0,
      },
      [],
    ]
  }

  const pctF = mujeres / totalBinarios
  const cumple = Math.abs(pctF - 0.5) <= UMBRAL_TRANSVERSAL_TOLERANCIA
  const pctCumplimiento = Math.max(0.0, 100.0 - Math.abs(pctF - 0.5) * 200)

  const incumplimientos = []
  if (!cumple) {
    const generoMinoritario = mujeres < hombres ? 'F' : 'M'
    const delta = Math.ceil(Math.abs(hombres - mujeres) / 2)
    incumplimientos.push({
      tipo: 'PARIDAD_TRANSVERSAL',
      descripcion:
        `Distribución actual: ${mujeres} mujeres (${(pctF * 100).toFixed(1)}%) y ` +
        `${hombres} hombres (${((1 - pctF) * 100).toFixed(1)}%). ` +
        `Se requiere 50% ±${(UMBRAL_TRANSVERSAL_TOLERANCIA * 100).toFixed(0)}%.`,
      candidatos_afectados: [],
      articulo: ART_PARIDAD_TRANSVERSAL,
      sugerencia:
        `Agregar o sustituir al menos ${delta} candidatura(s) de género ` +
        `'${generoMinoritario}' para alcanzar la paridad del 50% (tolerancia ±5%).`,
    })
  }

  const criterio = {
    cumple,
    porcentaje_cumplimiento: round2(pctCumplimiento),
    articulo: ART_PARIDAD_TRANSVERSAL,
    descripcion:
      'El total de candidaturas debe ser 50% mujeres y 50% hombres, con tolerancia del ±5%.',
    mujeres,
    hombres,
    porcentaje_mujeres: round2(pctF * 100),
  }
  return [criterio, incumplimientos]
}

// ---------------------------------------------------------------------------
// Reglas 4–7 — Acciones Afirmativas (Art. 14 LGIPE)
// ---------------------------------------------------------------------------
function accionAfirmativa(candidaturas, campo, umbral, nombreAccion, tipoIncumplimiento, descripcionCriterio) {
  const total = candidaturas.length
  const registrado = candidaturas.filter((c) => Boolean(c[campo])).length
  const requerido = Math.ceil(total * umbral)

  const cumple = registrado >= requerido
  const pct = Math.min(100.0, requerido > 0 ? (registrado / requerido) * 100 : 100.0)

  const incumplimientos = []
  if (!cumple) {
    const faltantes = requerido - registrado
    incumplimientos.push({
      tipo: tipoIncumplimiento,
      descripcion:
        `Se registraron ${registrado} candidatura(s) de ${nombreAccion}, ` +
        `pero se requieren al menos ${requerido} ` +
        `(${(umbral * 100).toFixed(0)}% de ${total} candidaturas).`,
      candidatos_afectados: [],
      articulo: ART_ACCION,
      sugerencia: `Registrar al menos ${faltantes} candidatura(s) adicional(es) de ${nombreAccion}.`,
    })
  }

  const criterio = {
    cumple,
    porcentaje_cumplimiento: round2(pct),
    articulo: ART_ACCION,
    descripcion: descripcionCriterio,
    registrado,
    requerido,
  }
  return [criterio, incumplimientos]
}

function accionJuventud(candidaturas) {
  const total = candidaturas.length
  const registrado = candidaturas.filter((c) => {
    const anio = anioNacimiento(c)
    return anio >= ANIO_JUVENTUD_MIN && anio <= ANIO_JUVENTUD_MAX
  }).length
  const requerido = Math.ceil(total * UMBRAL_JUVENTUD)
  const cumple = registrado >= requerido
  const pct = Math.min(100.0, requerido > 0 ? (registrado / requerido) * 100 : 100.0)

  const incumplimientos = []
  if (!cumple) {
    const faltantes = requerido - registrado
    incumplimientos.push({
      tipo: 'ACCION_JUVENTUD',
      descripcion:
        `Se registraron ${registrado} candidatura(s) de personas jóvenes ` +
        `(nacidas ${ANIO_JUVENTUD_MIN}-${ANIO_JUVENTUD_MAX}), ` +
        `pero se requieren al menos ${requerido} ` +
        `(${(UMBRAL_JUVENTUD * 100).toFixed(0)}% de ${total} candidaturas).`,
      candidatos_afectados: [],
      articulo: ART_ACCION,
      sugerencia:
        `Registrar al menos ${faltantes} candidatura(s) adicional(es) ` +
        `de personas nacidas entre ${ANIO_JUVENTUD_MIN} y ${ANIO_JUVENTUD_MAX}.`,
    })
  }

  const criterio = {
    cumple,
    porcentaje_cumplimiento: round2(pct),
    articulo: ART_ACCION,
    descripcion:
      `Al menos ${(UMBRAL_JUVENTUD * 100).toFixed(0)}% de las candidaturas deben ser de personas ` +
      `jóvenes (nacidas entre ${ANIO_JUVENTUD_MIN} y ${ANIO_JUVENTUD_MAX}).`,
    registrado,
    requerido,
  }
  return [criterio, incumplimientos]
}

// ---------------------------------------------------------------------------
// Función principal
// ---------------------------------------------------------------------------
export function validar(candidaturas) {
  const partido = candidaturas.length ? candidaturas[0].partido : 'DESCONOCIDO'
  const total = candidaturas.length

  const [cHorizontal, vHorizontal] = paridadHorizontal(candidaturas)
  const [cVertical, vVertical] = paridadVertical(candidaturas)
  const [cTransversal, vTransversal] = paridadTransversal(candidaturas)

  const [cIndigena, vIndigena] = accionAfirmativa(
    candidaturas, 'indigena', UMBRAL_INDIGENA, 'personas indígenas', 'ACCION_INDIGENA',
    `Al menos ${(UMBRAL_INDIGENA * 100).toFixed(0)}% de las candidaturas deben ser de personas indígenas.`,
  )
  const [cDiscapacidad, vDiscapacidad] = accionAfirmativa(
    candidaturas, 'discapacidad', UMBRAL_DISCAPACIDAD, 'personas con discapacidad', 'ACCION_DISCAPACIDAD',
    `Al menos ${(UMBRAL_DISCAPACIDAD * 100).toFixed(0)}% de las candidaturas deben ser de personas con discapacidad.`,
  )
  const [cJuventud, vJuventud] = accionJuventud(candidaturas)
  const [cLgbtq, vLgbtq] = accionAfirmativa(
    candidaturas, 'lgbtq', UMBRAL_LGBTQ, 'personas LGBTQ+', 'ACCION_LGBTQ',
    `Al menos ${(UMBRAL_LGBTQ * 100).toFixed(0)}% de las candidaturas deben ser de personas LGBTQ+.`,
  )

  const fallaParidad = !cHorizontal.cumple || !cVertical.cumple || !cTransversal.cumple
  const fallaAfirmativa =
    !cIndigena.cumple || !cDiscapacidad.cumple || !cJuventud.cumple || !cLgbtq.cumple

  let resultadoGlobal = 'SUCCESS'
  if (fallaParidad) resultadoGlobal = 'DANGER'
  else if (fallaAfirmativa) resultadoGlobal = 'WARNING'

  const incumplimientos = [
    ...vHorizontal, ...vVertical, ...vTransversal,
    ...vIndigena, ...vDiscapacidad, ...vJuventud, ...vLgbtq,
  ]

  return {
    partido,
    total_candidaturas: total,
    resultado_global: resultadoGlobal,
    criterios: {
      paridad_horizontal: cHorizontal,
      paridad_vertical: cVertical,
      paridad_transversal: cTransversal,
      accion_indigena: cIndigena,
      accion_discapacidad: cDiscapacidad,
      accion_juventud: cJuventud,
      accion_lgbtq: cLgbtq,
    },
    incumplimientos,
  }
}
